man/krisplookup.1
.TH KRISPLOOKUP 1 "11 Jul 2016"

.SH NAME
.BI krisplookup
\- Obtain Country, ISP information of any IP address.

.SH SYNOPSIS
.BI "krisplookup \-[fhin] <" ipaddress "|" hostname ">"

.SH DESCRIPTION
krisplookup is utility to get information of Country and ISP
about specify ip address.
.PP
For example,
.PP
.BI "krisplookup " oops.org
.PP
command prints follow informations.
.PP
.BI "oops.org (210.124.122.26): (주)데이콤 (BORANET)"
.br
.BI "SUBNET : 255.255.0.0"
.br
.BI "NETWORK: 210.124.0.0"
.br
.BI "BCAST  : 210.124.255.255"
.br
.BI "NATION : Korea, Republic of (KR)"
.br
.PP
.SH OPTIONS
.IP "-f"
Specifies user database. Basically krisp database is been hard cording.
If you want to open other database file as krisp, using -f option.
.IP "-h"
Print help messages.
.IP "-i"
Print only isp code.
.IP "-n"
Print only country code.

.SH AUTHORS
JoungKyun.Kim <https://joungkyun.github.com>

.SH BUGS
Report to https://github.com/Joungkyun/libkrisp/issues

.SH COPYRIGHT
Copyright (c) 2016 JoungKyun.Kim <http://oops.org>

This api is follows
.BI LGPL 2.1


Changes
Changelog


* 3.1.6
Tue, Feb 17 2016 JoungKyun.Kim <http://oops.org>
	- start 3.1.6


* 3.1.5
Tue, Feb 16 2016 JoungKyun.Kim <http://oops.org>
	- fixed access uninitialized variable

Sat, Feb 15 2013 JoungKyun.Kim <http://oops.org>
	- fixed libiconv link on cygwin

Fri, Feb 14 2013 JoungKyun.Kim <http://oops.org>
	- support windows native binary
      use MinGW32 with Code::Blocks workspace
	- fixed wrong db close when database file not found

Tue, Feb 11 2013 JoungKyun.Kim <http://oops.org>
	- fixed build error and shared library on cygwin

Fri, Oct 25 2013 JoungKyun.Kim <http://oops.org>
	- start 3.1.5 tree


* 3.1.4
Fri, Oct 25 2013 JoungKyun.Kim <http://oops.org>
	- release 3.1.4
	- fixed opening the wrong path of database
	- add kr_range API for ISP or Country

Thu, Nov  1 2012 JoungKyun.Kim <http://oops.org>
	- fixed libipcalc link

Tue, Sep 18 2012 JoungKyun.Kim <http://oops.org>
	- fixed cygwin compile environments


* 3.1.3
Mon, Feb 27 2012 JoungKyun.Kim <http://oops.org>
	- release 3.1.3
	- fixed missing krversion.h

Mon, May  2 2011 JoungKyun.Kim <http://oops.org>
	- add 'extern C' for C plus plus

Sat, Apr 30 2011 JoungKyun.Kim <http://oops.org>
	- enhanced krisp.m4

Fri, Apr 22 2011 JoungKyun.Kim <http://oops.org>
	- changed build environment with automake
	- revert gcc visibility


* 3.1.2
Fri, Jan 14 2011 JoungKyun.Kim <http://oops.org>
	- release 3.1.2
	- fixed segfault cause of miss casting

Tue, Jan 12 2011 JoungKyun.Kim <http://oops.org>
	- add english man pages
	- debianized

Mon, Nov  8 2010 JoungKyun.Kim <http://oops.org>
	- fixed build environments on other OS (ulong type, make command)
	- 3.1.2 dev branch start


* 3.1.1
Mon, Sep 27 2010 JoungKyun.Kim <http://oops.org>
	- release 3.1.1

Fri, Sep 24 2010 JoungKyun.Kim <http://oops.org>
	- add journal_mode=OFF option on sqlite3 database handle
	- add _REENTRANT compile flag
	- changed default value of db_time_stamp_interval from 3600 sec to 0 sec

Mon, Sep 20 2010 JoungKyun.Kim <http://oops.org>
	- initialize db handler to NULL when database connect failed on kr_open


* 3.1.0
Fri, Sep 10 2010 JoungKyun.Kim <http://oops.org>
	- changed kr_close api proto type
	  kr_close (KR_API *) => kr_close (KR_API **)
	- reconnect database when changed mtime of database
	  add 4 members in KR_NET struct
	  database
	  db_time_stamp_interval
	  db_time_stamp
	  db_stamp_checked


* 3.0.3

Fri, Sep 10 2010 JoungKyun.Kim <http://oops.org>
	- fixed segfault kr_search(_ex) when KR_NET structure is null

Thu, Aug 12 2010 JoungKyun.Kim <http://oops.org>
	- fixed typo


* 3.0.2

Sun, Aug  8 2010 JoungKyun.Kim <http://oops.org>
	- release 3.0.2
	- link with libipcalc
	- remove ip2long, netcalc that forked libipcalc
	- add --with-compat build option for makeing compat library
	- add krisp.pc that is pkg-config configuration file


* 3.0.1

Tue, Aug  3 2010 JoungKyun.Kim <http://oops.org>
	- release 3.0.1
	- moved header files to %includedir/krisp
	- removed dependency library on --libs and, add --link for dependency
	  library on krisp-config


* 3.0.0

Fri, Jul 16 2010 JoungKyun.Kim <http://oops.org>
	- release 3.0.0
	- update database on july

Fri, Jul  2 2010 JoungKyun.Kim <http://oops.org>
	- enhanced build environments

Wed, Jun 30 2010 JoungKyun.Kim <http://oops.org>
	- add kr_long2ip_r api for thread safe
	- changed mutex lock per db constructor

Tue, Jun 29 2010 JoungKyun.Kim <http://oops.org>
	- rewrite _kr_valid_ip_address and change proto type

Sun, Jun 27 2010 JoungKyun.Kim <http://oops.org>
	- initialize verbose member of KRNET_API(_ex) on kr_search(_ex) API

Fri, Jun 25 2010 JoungKyun.Kim <http://oops.org>
	- rename internal api name
	- make local symbol
	- fix wrong caculation of netcalc when use FQDN instead of IP address

Thu, Jun 24 2010 JoungKyun.Kim <http://oops.org>
	- support multiple build (make -j4)
	- fixed dl link error

Fri, Jun 18 2010 JoungKyun.Kim <http://oops.org>
	- fixed sqlite3_open_v2 compatibility problems under sqlite 3.5.0
	- add bool type to change some APIs and variables and structure members
	- don't call kr_close api any more when failed kr_open

Wed, Jun 16 2010 JoungKyun.Kim <http://oops.org>
	- changed proto type of kr_open
	- add err member on kr_search/kr_search_ex for thread safe

Mon, Jun 14 2010 JoungKyun.Kim <http://oops.org>
	- add thread safe

Thu, Jun 10 2010 JoungKyun.Kim <http://oops.org>
	- add database file check on kr_open

Wed, Jun  9 2010 JoungKyun.Kim <http://oops.org>
	- fixed segfault on kr_open

Tue, Jun  8 2010 JoungKyun.Kim <http://oops.org>
	- init verbose member of KR_API to 0 on kr_open
	- add dummy member array size on RAW_KRNET_API
	- changed frontend api from function to define

Sat, May 22 2010 JoungKyun.Kim <http://oops.org>
	- changed query algoritm
	- changed database structure
	- support internal geoip/geoipisp database
	  Don't need geoip library any more
	  Support convert scripts
	- update SONAME to 2.0.0
	  changed usage of almost APIs




utils/krisplookup.c
/*
 * Copyright (c) 2016 JoungKyun.Kim <http://oops.org> all rights reserved
 *
 * This file is part of libkrisp
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU LESSER GENERAL PUBLIC LICENSE as published
 * by the Free Software Foundation; either version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU LESSER GENERAL PUBLIC LICENSE for more details.
 *
 * You should have received a copy of the GNU LESSER GENERAL PUBLIC LICENSE
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <krispapi.h>

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#define PNAME "krisplookup"

#ifdef HAVE_ICONV_H
#include <iconv.h>
#include <errno.h>
#define LOCALEALIAS "/usr/share/locale/locale.alias"
#define DB_CHARSET "UTF8"
#define FORCE_NOTHING 0
#define FORCE_UTF8 1
#define FORCE_EUCKR 2
#endif

#ifdef _WIN32
	#include <ws2tcpip.h>
	#define WSA_Cleanup WSACleanup()
#else
	#define WSA_Cleanup
#endif


#ifdef HAVE_GETOPT_LONG
static struct option long_options [] = { // {{{
	/* Options without arguments: */
	{ "help",       no_argument,       NULL, 'h' },
	{ "verbose",    no_argument,       NULL, 'v' },
	{ "onlyisp",    no_argument,       NULL, 'i' },
	{ "onlynation", no_argument,       NULL, 'n' },

	/* Options accepting an argument: */
	{ "datafile",   required_argument, NULL, 'f' },
	{ "range",		required_argument, NULL, 'r' },
	{ 0, 0, 0, 0 }
}; // }}}
#endif

#ifdef HAVE_ICONV_H
char * confirm_local_charset (void) { // {{{
	char *	lang_env = getenv ("LANG");
	char *	lcharset = strrchr (lang_env + 1, '.');
	char	buf[1024];
	char *	tmp;

	if ( lcharset == NULL ) {
		struct stat		f;
		FILE *			fp;
		char *			line;

		f.st_size = 0;
		if ( stat (LOCALEALIAS, &f) == -1 )
			return NULL;

		if ( (fp = fopen (LOCALEALIAS, "r")) == NULL )
			return NULL;

		while ( fgets (buf, 1024, fp) != NULL ) {
			if ( buf[0] < 97 && buf[0] > 122 )
				continue;

			if ( (line = strstr (buf, lang_env)) != NULL ) {
				if ( (tmp = strchr (buf, '.')) != NULL )
					return tmp + 1;
			}
		}

		return NULL;
	}

	if ( ! strcasecmp (".utf-8", lcharset) || ! strcasecmp (".utf8", lcharset) )
		return "UTF8";

	return lcharset + 1;
} // }}}
#endif

void usage (char *prog) { // {{{
	fprintf (stderr, "%s v%s: Resolved korea range ISP inforamtion\n", prog, krisp_version ());
	fprintf (stderr, "Usage: %s [option] ip-address\n", prog);
	fprintf (stderr, "Options:\n");
	fprintf (stderr, "         -f path, --datafile=path     set user define database file\n");
	fprintf (stderr, "         -h , --help                  print this message\n");
	fprintf (stderr, "         -i , --isp                   only print isp code\n");
	fprintf (stderr, "         -n , --nation                only print nation code\n");
	fprintf (stderr, "         -r , --range=[country|isp]   Print all range of Country or ISP\n");
	fprintf (stderr, "                                      about current IP\n\n");

	exit (1);
} // }}}

int main (int argc, char ** argv) {
	KR_API *		db;
	KRNET_API		isp;
	char *			ip;
	int				opt;
	char *			datafile = NULL;
	short			onlyisp = 0;
	short			onlynation = 0;
	short			printrange = 0;
	short			printtype  = KRISP_GET_COUNTRY;
	bool			verbose = false;
	char			err[1024];

#ifdef HAVE_GETOPT_LONG
	while ( (opt = getopt_long (argc, argv, "f:hinr:v", long_options, (int *) 0)) != EOF ) {
#else
	while ( (opt = getopt (argc, argv, "f:hinr:v")) != EOF ) {
#endif
		switch (opt) {
			case 'f' :
				datafile = optarg;
				break;
			case 'i' :
				if ( onlynation > 0 ) {
					fprintf (stderr, "ERROR: Can't use -n option and -i option togather\n");
					return 1;
				}
				onlyisp++;
				break;
			case 'n' :
				if ( onlyisp > 0 ) {
					fprintf (stderr, "ERROR: Can't use -n option and -i option togather\n");
					return 1;
				}
				onlynation++;
				break;
			case 'r' :
				printrange++;

				if ( ! strcasecmp ("country", optarg) )
					printtype = KRISP_GET_COUNTRY;
				else if ( ! strcasecmp ("isp", optarg) )
					printtype = KRISP_GET_ISP;
				else {
					fprintf (stderr, "ERROR: Wrong value of -r option\n");
					return 1;
				}
				break;
			case 'v' :
				set_true (verbose);
				break;
			default:
				usage (PNAME);
		}
	}

	if ( argc - optind < 1 || argc == 1 ) {
		usage (PNAME);
		return 1;
	}

#ifdef _WIN32
	{
		WORD wVerReq = MAKEWORD (2, 2); // Call WinSock 2.2
		WSADATA wsaData;
		int nErrStatus;

		if ( (nErrStatus = WSAStartup (wVerReq, &wsaData)) != 0 ) {
			fprintf (stderr, "Error: Failed initialize WSAStartup\n");
			return 1;
		}
	}
#endif

	/* database open */
	if ( kr_open (&db, datafile, err) == false ) {
		fprintf (stderr, "ERROR Connect: %s\n", err);
		WSA_Cleanup;
		return 1;
	}

	db->db_time_stamp_interval = 0;

	isp.verbose = verbose;
	db->verbose = verbose;
	ip = argv[optind];

	SAFECPY_256 (isp.ip, ip);
	if ( kr_search (&isp, db) ) {
		fprintf (stderr, "ERROR: %s\n", isp.err);
		kr_close (&db);
		WSA_Cleanup;
		return 1;
	}

	if ( verbose )
		fprintf (stderr, "\n");

	if ( strlen (isp.err) )
		printf ("Error: %s\n", isp.err);

	if ( onlyisp ) {
		printf ("%s\n", isp.icode);
	} else if ( onlynation ) {
		printf ("%s\n", isp.ccode);
	} else {
#ifdef HAVE_ICONV_H
		char *	lcharset;
		iconv_t	cd;
		char *	ispname, * to;
		size_t	flen, tlen;
		char *	srcname = (char *) isp.iname;

#if defined _WIN32
		lcharset = "CP949";
#else
		lcharset = confirm_local_charset ();
#endif
		flen = strlen (srcname);

		if ( lcharset == NULL || ! strcmp (DB_CHARSET, lcharset) ) {
			tlen = 1;
			cd = (iconv_t)(-1);
		} else {
			tlen = ! strcmp (lcharset, "UTF8") ? flen * 4 + 1 : flen + 1;
			cd = iconv_open (lcharset, DB_CHARSET);
		}

		if ( cd == (iconv_t)(-1) ) {
			ispname = strdup (isp.iname);
			goto noconvert;
		}

		if ( (ispname = (char *) malloc (sizeof (char) * tlen)) == NULL ) {
			ispname = strdup (isp.iname);
			goto noconvert;
		}
		memset (ispname, 0, sizeof (char) * tlen);
		to = ispname;

		iconv (cd, &srcname, &flen, &to, &tlen);
		switch ( errno ) {
			case E2BIG :
			case EILSEQ :
			case EINVAL :
				strcpy (ispname, isp.iname);
				break;
		}

noconvert:
		printf ("%s (%s): %s (%s)\n", ip, isp.ip, ispname, isp.icode);
#else
		printf ("%s (%s): %s (%s)\n", ip, isp.ip, isp.iname, isp.icode);
#endif

#ifdef HAVE_ICONV_H
		if ( cd != (iconv_t)(-1) )
			iconv_close (cd);
		SAFEFREE (ispname);
#endif

		printf ("SUBNET    : %s\n", long2ip (isp.netmask));
		printf (
				"NETWORK   : %s\n",
				long2ip (network (isp.start, isp.netmask))
		);
		printf (
				"BROADCAST : %s\n",
				long2ip (broadcast(isp.start, isp.netmask))
		);
		printf ("DB RANGE  : %s - ", long2ip (isp.start));
		printf ("%s\n", long2ip (isp.end));
		printf ("NATION    : %s (%s)\n", isp.cname, isp.ccode);
	}

	if ( printrange ) {
		KRNET_REQ_RANGE range;
		int i;
		ulong mask;
		char start[16];
		char end[16];

		range.code = printtype;
		range.verbose = verbose;
		range.ranges = NULL;
		range.count = 0;

		SAFECPY_1024 (
			range.data,
			(printtype == KRISP_GET_COUNTRY) ? isp.ccode : isp.iname
		);

		if ( kr_range (&range, db) ) {
			fprintf (stderr, "ERROR: %s\n", range.err);
			SAFEFREE (range.ranges);
			kr_close (&db);
			WSA_Cleanup;
			return 1;
		}

		printf (
			"\nRange Traget: %s (%d line%s)\n\n",
			range.data,
			range.count,
			range.count ? "s" : ""
		);

		for ( i=0; i<range.count; i++ ) {
			long2ip_r (range.ranges[i].start, start);
			mask = guess_netmask (range.ranges[i].start, range.ranges[i].end);

			if ( network (range.ranges[i].start, mask) == range.ranges[i].start )
				printf ("%s/%d\n", start, long2prefix (mask));
			else
				printf ("%s %s\n", start, long2ip_r (range.ranges[i].end, end));
		}

		SAFEFREE (range.ranges);
		printf ("\n");
	}

	/* database close */
	kr_close (&db);
	WSA_Cleanup;

	return 0;
}

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */
